ethereum/eth/src/main/java/tech/pegasys/pantheon/ethereum/eth/sync/tasks/ParallelDownloadHeadersTask.java

```diff
@@ -0,0 +1,96 @@
+ */
+package tech.pegasys.pantheon.ethereum.eth.sync.tasks;
+
+import tech.pegasys.pantheon.ethereum.ProtocolContext;
+import tech.pegasys.pantheon.ethereum.core.BlockHeader;
+import tech.pegasys.pantheon.ethereum.eth.manager.AbstractPipelinedPeerTask;
+import tech.pegasys.pantheon.ethereum.eth.manager.EthContext;
+import tech.pegasys.pantheon.ethereum.eth.manager.EthPeer;
+import tech.pegasys.pantheon.ethereum.mainnet.ProtocolSchedule;
+import tech.pegasys.pantheon.metrics.LabelledMetric;
+import tech.pegasys.pantheon.metrics.OperationTimer;
+
+import java.util.List;
+import java.util.Optional;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+
+import com.google.common.collect.Lists;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+public class ParallelDownloadHeadersTask<C>
+    extends AbstractPipelinedPeerTask<BlockHeader, List<BlockHeader>> {
+  private static final Logger LOG = LogManager.getLogger();
+
+  private final ProtocolSchedule<C> protocolSchedule;
+  private final ProtocolContext<C> protocolContext;
+
+  ParallelDownloadHeadersTask(
+      final BlockingQueue<BlockHeader> inboundQueue,
+      final int outboundBacklogSize,
+      final ProtocolSchedule<C> protocolSchedule,
+      final ProtocolContext<C> protocolContext,
+      final EthContext ethContext,
+      final LabelledMetric<OperationTimer> ethTasksTimer) {
+    super(inboundQueue, outboundBacklogSize, ethContext, ethTasksTimer);
+
+    this.protocolSchedule = protocolSchedule;
+    this.protocolContext = protocolContext;
+  }
+
+  @Override
+  protected Optional<List<BlockHeader>> processStep(
+      final BlockHeader nextCheckpointHeader,
+      final Optional<BlockHeader> previousCheckpointHeader,
+      final EthPeer peer) {
+    if (!previousCheckpointHeader.isPresent()) {
+      return Optional.empty();
+    }
+    final int segmentLength =
+        (int) (nextCheckpointHeader.getNumber() - previousCheckpointHeader.get().getNumber()) - 1;
+    // body
+    LOG.trace(
+        "Requesting download of {} blocks ending at {}",
+        segmentLength,
+        nextCheckpointHeader.getHash());
+    final DownloadHeaderSequenceTask<C> downloadTask =
+        DownloadHeaderSequenceTask.endingAtHeader(
+            protocolSchedule,
+            protocolContext,
+            ethContext,
+            nextCheckpointHeader,
+            segmentLength,
+            ethTasksTimer);
+    downloadTask.assignPeer(peer);
+    final CompletableFuture<List<BlockHeader>> headerFuture = executeSubTask(downloadTask::run);
+
+    final List<BlockHeader> headers = Lists.newArrayList(previousCheckpointHeader.get());
+    try {
+      headers.addAll(headerFuture.get());
+    } catch (final InterruptedException | ExecutionException e) {
+      result.get().completeExceptionally(e);
+      return Optional.empty();
+    }
+    headers.add(nextCheckpointHeader);
```

Comment:
Doesn't `DownloadHeaderSequenceTask.endingAtHeader` include `nextCheckpointHeader` as the last thing it retrieves?  So would this wind up adding it twice?

Reply:
We do need it: https://jenkins.pegasys.tech/job/Pantheon/view/change-requests/job/PR-774/17/#showFailuresLink